src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -0,0 +1,288 @@
+	Oid colocatedRelationId = InvalidOid;
+	Const *partitionValue = NULL;
+	Datum partitionValueDatum = 0;
+	ShardInterval *shardInterval = NULL;
+	List *placementList = NIL;
+	DistTableCacheEntry *distTable = NULL;
+	Var *partitionColumn = NULL;
+	ShardPlacement *placement = NULL;
+	WorkerNode *workerNode = NULL;
+	StringInfo queryString = NULL;
+	Task *task = NULL;
+	Job *job = NULL;
+	DistributedPlan *distributedPlan = NULL;
+	int32 groupId = 0;
+
+	if (!CitusHasBeenLoaded() || !CheckCitusVersion(DEBUG4))
+	{
+		/* Citus is not ready to determine whether function is distributed */
+		return NULL;
+	}
+
+	groupId = GetLocalGroupId();
+	if (groupId != 0 || groupId == GROUP_ID_UPGRADING)
+	{
+		/* do not delegate from workers, or while upgrading */
+		return NULL;
+	}
+
+	if (query == NULL)
+	{
+		/* no query (mostly here to be defensive) */
+		return NULL;
+	}
+
+	joinTree = query->jointree;
+	if (joinTree == NULL)
+	{
+		/* no join tree (mostly here to be defensive) */
+		return NULL;
+	}
+
+	if (joinTree->quals != NULL)
+	{
+		/* query has a WHERE section */
+		return NULL;
+	}
+
+	if (joinTree->fromlist != NIL)
+	{
+		/* query has a FROM section */
+#if PG_VERSION_NUM >= 120000
+
+		/* in pg12 empty FROMs are represented with an RTE_RESULT */
+		if (list_length(joinTree->fromlist) == 1)
+		{
+			RangeTblRef *reference = linitial(joinTree->fromlist);
+			RangeTblEntry *rtentry = rt_fetch(reference->rtindex, query->rtable);
+			if (rtentry->rtekind != RTE_RESULT)
+			{
+				return NULL;
+			}
+		}
+		else
+		{
+			return NULL;
+		}
+#else
+		return NULL;
+#endif
+	}
+
+	targetList = query->targetList;
+	if (list_length(query->targetList) != 1)
+	{
+		/* multiple target list items */
+		return NULL;
+	}
+
+	targetEntry = (TargetEntry *) linitial(targetList);
+	if (!IsA(targetEntry->expr, FuncExpr))
+	{
+		/* target list item is not a function call */
+		ereport(DEBUG4, (errmsg("query is not a function call")));
+		return NULL;
+	}
+
+	funcExpr = (FuncExpr *) targetEntry->expr;
+	procedure = LookupDistObjectCacheEntry(ProcedureRelationId, funcExpr->funcid, 0);
+	if (procedure == NULL)
+	{
+		/* not a distributed function call */
+		ereport(DEBUG4, (errmsg("function is not distributed")));
+		return NULL;
+	}
+
+	if (IsMultiStatementTransaction())
+	{
+		/* cannot delegate function calls in a multi-statement transaction */
+		ereport(DEBUG1, (errmsg("not pushing down function calls in "
+								"a multi-statement transaction")));
+		return NULL;
+	}
+
+	if (procedure->distributionArgIndex < 0 ||
+		procedure->distributionArgIndex >= list_length(funcExpr->args))
+	{
+		ereport(DEBUG1, (errmsg("function call does not have a distribution argument")));
+		return NULL;
+	}
+
+	if (contain_volatile_functions((Node *) funcExpr->args))
+	{
+		ereport(DEBUG1, (errmsg("arguments in a distributed function must "
+								"be constant expressions")));
+		return NULL;
+	}
+
+	if (expression_tree_walker((Node *) funcExpr->args, contain_param_walker, NULL))
```

Comment:
Do we ever need to check for `PARAM_EXEC` as well? Can you double check?

Reply:
Do you mean we should just not push down for `PARAM_EXEC` and push-down other kinds of params? (paramkind is `PARAM_EXEC` in our subquery test).

I also added a prepared statement test. It seems that by the time the code arrives here parameters have been replaced.

So it seems that we can just reject any param type for now as we are doing now.